walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,97 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
 
 public final class FindMeetingQuery {
-  public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
-  }
+    public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
+        Collection<String> mandatoryAttendees = request.getAttendees();
+        Collection<String> allAttendees = request.getAllAttendees();
+        long duration = request.getDuration();
+
+        Collection<Event> mandatoryAttendeeEvents = getEventsForAttendees(mandatoryAttendees, events);
+        ArrayList<TimeRange> mandatoryAttendeesSlots = getPossibleSlots(mandatoryAttendeeEvents, mandatoryAttendees, duration);
+
+        Collection<Event> allAttendeeEvents = getEventsForAttendees(allAttendees, events);
+        ArrayList<TimeRange> allAttendeeSlots = getPossibleSlots(allAttendeeEvents, allAttendees, duration);
+
+        if (allAttendeeSlots.size() == 0 && mandatoryAttendees.size() != 0) {
+            return mandatoryAttendeesSlots;
+        }
+        else {
+            return allAttendeeSlots;
+        }
+
+    }
+
+    public Collection<Event> getEventsForAttendees(Collection<String> attendees, Collection<Event> events) {
+        ArrayList<Event> filteredEvents = new ArrayList<>();;
+        for (Event event : events) {
+            Set<String> eventAttendees = event.getAttendees();
+            for (String attendee : attendees) {
+                if (eventAttendees.contains(attendee)) {
+                    filteredEvents.add(event);
+                    break;
+                }
+            }
+        }
+        return filteredEvents;
```

Comment:
If you want, you can make this more concise with streams (up to you if you think it's more readable):
```
return events.stream()
    .filter(event -> !Collections.disjoint(attendees, event.getAttendees()))
    .collect(toList());
```

Reply:
I'm not familiar with streams so I don't want to throw this in just yet, but I'll do more research to understand this code.